### ISSUE TITLE
Limit order amount precision

### ISSUE THREAD

**Issue by amjad1210** [knowm/XChange#2497]
Issue #1544 talked about having the quantity_scale data for a pair which would be very useful when setting the limit order amount precise to the scale. Currently most of my orders are being rejected from the exchange API due to not setting a precise amount. @timmolter  @npomfret should i go ahead and create a PR with the suggestion or was there a reason this wasn't added maybe.

**Comment by timmolter** [comment]
Go for it. No one else got to it yet, that's why it's not done.

**Comment by perseus93** [comment]
I am actually running into the same issue, great to see you're already on top of it!
Would you have a (rough) ETA on your pull request? It would help me with prioritizing other tasks in my little project.

Additionally if you need any support drop me a message and I can see what I can do!

**Comment by amjad1210** [comment]
Hey @perseus93 while trying to solve the issue i figured out that we have the price scale in 2 separate places.

CurrencyMetaData and CurrencyPairMetaData

CurrencyPairMetaData should have the price scale for the pair e.g. 8 for a pair like ETH/BTC meaning 1 ETH will give you something like 0.40000000 BTC to your wallet.

CurrencyMetaData price scale should be the trade amount precision and withdrawal amount precision amount from an exchange and should be different to the scale set in CurrencyPairMetaData.

The issue is in some exchanges it has been setup where the price scale is using the CurrencyPairMetaData price scale in both locations. The best way to solve it is the rectify the issue on the exchange repository and create a PR hope this help.

**Comment by amjad1210** [comment]
If still needed i did commit the ability to support amount_scale but reverted it after finding out the issue above.

https://github.com/knowm/XChange/commit/d061731a6f0445418b523c4d85ccfb9f7c5f66c3

You can add this back in :)

@timmolter If you could shed some light on this it would help as i may be wrong.

**Comment by timmolter** [comment]
@amjad1210 can you show an example of a rejected limit order and an accepted limit order?

**Comment by perseus93** [comment]
@amjad1210 I don't think the price scale in the CurrencyPairMetadata is the same as the quantity scale.

If I understand your explanation correctly, the price scale in the CurrencyPairMetadata defines the precision of the amount obtained after executing an order.

What I need is the precision of the order that I can place, e.g. can I submit a LimitOrder with an original amount defined in 8 decimal places or only 2 decimal places.

In short, how I see it:
Price scale: how precise one can set the price of an order.
Quantity scale: how precise one can set the order volume of an order.

Disclaimer: I could be entirely incorrect, but this is how I interpret the api of the XChange project.

@timmolter I can't give you a concrete example but to clarify the problem that I am running into is the following:

At the end of the line my program creates a LimitOrder, the values for that limit order are computed by me.
If I would submit an order on an exchange e.g. EUR/BTC and its precision is limited to cents, submitting an order of a volume of 12.345 euro would result in a rejected order. (the exchange does not accept the precision of that order volume).

**Comment by amjad1210** [comment]
@perseus93 This does seem to be an issue from further investigation. I've been using the CurrenyMetaData scale located in org.knowm.xchange.dto.meta. Could someone possibly clarify what the scale field in CurrenyMetaData class is for? I've been using it for the limit amount/withdrawal amount precision scaling and discovered it works with some and not others.

**Comment by amjad1210** [comment]
So from my understanding.

CurrenyMetaData scale is the precision amount the exchange wallet holds the coin balance up to?

CurrencyPairMetadata scale is possibly the limit price precision?

So we're missing a concrete implementation for amount scale across all exchanges. 

@timmolter Any chance you could clarify the difference between CurrenyMetaData scale and CurrenyMetaData price scale. In conclusion the amount scale needs to be implemented either way so i'll make a start.

**Comment by timmolter** [comment]
@amjad1210 Yes you are correct. For CurrencyPairMetadata you never want to place a limit order with a price with higher scale (precision) than the exchange allows. Looks like we need an amount scale in CurrencyPairMetadata as well.

**Comment by ThomasWick** [comment]
On a similar note i've notice that the amount scale precision is hard coded to 2 in the ACXTradeService placeOrder() method.

Is there a specific reason for this?

**Comment by timmolter** [comment]
@ThomasWick There's no reason for it.

**Comment by ThomasWick** [comment]
@timmolter i've removed the hard coded precision from ACX. This would allow responsibility for precision to be with CurrencyPairMetaData only (for ACX).

Can create a PR if that's ok?

**Comment by timmolter** [comment]
yes.

**Comment by amjad1210** [comment]
I know this request has been opened for a while and I've finally figured out what was missing and where.

CurrenyMetaData: scale =  the precision amount the exchange wallet holds the coin balance up to.
CurrencyPairMetadata:  priceScale = the quantity precision scale for the counter coin.

I've added a new variable to CurrencyPairMetadata called baseScale which is the quantity precision scale for the base coin.

Will open a PR with the changes.
